File: web/lib/format.ts
/**
 * Formatting and wording, in one place.
 *
 * The labels matter more here than in most systems. Everything below is read
 * by a worried parent on a phone, and the difference between "الحالة: غياب"
 * and "ما إجا ع الجلسة" is the difference between a case file and a letter.
 * The keys stay in English and stay stable — they are what the API and the
 * filters match on — so the wording can be argued about without touching any
 * logic.
 */

export function number(value: string | number | null | undefined): string {
  return Number(value ?? 0).toLocaleString("en-US");
}

export function percent(value: number | null | undefined): string {
  if (value === null || value === undefined) return "—";
  return `${Math.round(value)}%`;
}

const MONTHS = [
  "يناير", "فبراير", "مارس", "أبريل", "مايو", "يونيو",
  "يوليو", "أغسطس", "سبتمبر", "أكتوبر", "نوفمبر", "ديسمبر",
];

const DAYS = ["الأحد", "الإثنين", "الثلاثاء", "الأربعاء", "الخميس", "الجمعة", "السبت"];

function parse(value: string | null | undefined): Date | null {
  if (!value) return null;
  const d = new Date(value.replace(" ", "T"));
  return Number.isNaN(d.getTime()) ? null : d;
}

/** 23 أغسطس 2026 */
export function date(value: string | null | undefined): string {
  const d = parse(value);
  if (!d) return "—";
  return `${d.getDate()} ${MONTHS[d.getMonth()]} ${d.getFullYear()}`;
}

/** الأحد 23 أغسطس — for a diary, where the weekday is what is scanned for. */
export function dayDate(value: string | null | undefined): string {
  const d = parse(value);
  if (!d) return "—";
  return `${DAYS[d.getDay()]} ${d.getDate()} ${MONTHS[d.getMonth()]}`;
}

/** 23 أغسطس، 10:30 */
export function dateTime(value: string | null | undefined): string {
  const d = parse(value);
  if (!d) return "—";
  return `${d.getDate()} ${MONTHS[d.getMonth()]}، ${time(value)}`;
}

/** 10:30 */
export function time(value: string | null | undefined): string {
  const d = parse(value);
  if (!d) return "—";
  return `${String(d.getHours()).padStart(2, "0")}:${String(d.getMinutes()).padStart(2, "0")}`;
}

/**
 * "قبل ٣ أيام" / "بعد ساعتين".
 *
 * Arabic counts in three forms — one, two, and many — and getting it wrong
 * ("قبل 2 يوم") is the kind of thing that makes software read as foreign.
 */
export function relative(value: string | null | undefined): string {
  const d = parse(value);
  if (!d) return "—";

  const diff = d.getTime() - Date.now();
  const past = diff < 0;
  const minutes = Math.round(Math.abs(diff) / 60000);

  const phrase = (n: number, one: string, two: string, few: string, many: string) => {
    if (n === 1) return one;
    if (n === 2) return two;
    if (n <= 10) return `${n} ${few}`;
    return `${n} ${many}`;
  };

  let unit: string;

  if (minutes < 1) return "الآن";
  else if (minutes < 60) unit = phrase(minutes, "دقيقة", "دقيقتين", "دقائق", "دقيقة");
  else if (minutes < 1440) {
    const hours = Math.round(minutes / 60);
    unit = phrase(hours, "ساعة", "ساعتين", "ساعات", "ساعة");
  } else if (minutes < 43200) {
    const days = Math.round(minutes / 1440);
    unit = phrase(days, "يوم", "يومين", "أيام", "يوماً");
  } else {
    const months = Math.round(minutes / 43200);
    unit = phrase(months, "شهر", "شهرين", "شهور", "شهراً");
  }

  return past ? `قبل ${unit}` : `بعد ${unit}`;
}

/* -------------------------------------------------------------------------
 * The prompt scale
 * ---------------------------------------------------------------------- */

/**
 * The five steps, worded for whoever is reading them.
 *
 * A specialist scoring a goal and a mother reading the score need the same
 * five buckets to mean the same five things — so there is one set of words,
 * not a clinical set and a friendly set that drift apart.
 */
export const LEVEL_LABELS: Record<number, string> = {
  4: "مستقل",
  3: "تلميح لفظي",
  2: "مساعدة جزئية",
  1: "مساعدة كاملة",
  0: "رفض / ما استجاب",
};

/** What each step means, so two specialists score the same child the same. */
export const LEVEL_HINTS: Record<number, string> = {
  4: "عملها لحاله بدون أي تذكير.",
  3: "عملها لحاله بعد ما ذكّرناه بالكلام أو بالإشارة.",
  2: "بدأ لحاله واحتاج مساعدة بالنص، أو لمسة توجيه.",
  1: "عملناها معه يداً بيد من أولها لآخرها.",
  0: "ما قبل يجرّب، أو ما استجاب إطلاقاً.",
};

/** The validated ordinal ramp — see globals.css for what was measured. */
export function levelColor(level: number): string {
  return `var(--viz-level-${Math.max(0, Math.min(4, level))})`;
}

/**
 * The text colour to put ON a level swatch.
 *
 * The ramp's two lightest steps take dark ink; the rest take white. Computed
 * from the step index rather than eyeballed per component, so a swatch is
 * never white-on-pale anywhere in the app.
 */
export function levelInk(level: number): string {
  return level <= 1 ? "#04302b" : "#ffffff";
}

/* -------------------------------------------------------------------------
 * Session state
 * ---------------------------------------------------------------------- */

export const SESSION_LABELS: Record<string, string> = {
  scheduled: "مجدولة",
  held: "تمّت",
  absent: "ما إجا",
  excused: "غياب بعذر",
  cancelled: "ملغية",
};

export const SESSION_TONES: Record<string, string> = {
  scheduled: "bg-line/60 text-ink",
  held: "bg-good-soft text-good",
  absent: "bg-bad-soft text-bad",
  excused: "bg-warn-soft text-warn",
  cancelled: "bg-line/50 text-muted",
};

export const REPORT_LABELS: Record<string, string> = {
  draft: "مسوّدة",
  published: "وصل الأهل",
};

export const REPORT_TONES: Record<string, string> = {
  draft: "bg-warn-soft text-warn",
  published: "bg-good-soft text-good",
};

export const MOOD_LABELS: Record<string, string> = {
  calm: "هادئ",
  happy: "مبسوط",
  tired: "تعبان",
  agitated: "متوتّر",
  crying: "بيبكي",
  resistant: "رافض",
};

export const MOOD_FACES: Record<string, string> = {
  calm: "😌",
  happy: "😄",
  tired: "😪",
  agitated: "😣",
  crying: "😢",
  resistant: "😤",
};

export const SPECIALTY_LABELS: Record<string, string> = {
  speech: "النطق واللغة",
  occupational: "العلاج الوظيفي",
  behavioral: "تعديل السلوك",
  special_ed: "التربية الخاصة",
};

export const SPECIALTY_FULL: Record<string, string> = {
  speech: "النطق واللغة والتخاطب",
  occupational: "العلاج الوظيفي والتكامل الحسي",
  behavioral: "تعديل السلوك والتوحد",
  special_ed: "التربية الخاصة وصعوبات التعلّم",
};

export const GOAL_LABELS: Record<string, string> = {
  active: "شغّال",
  achieved: "تحقّق",
  paused: "موقوف",
  dropped: "أُلغي",
};

export const GOAL_TONES: Record<string, string> = {
  active: "bg-brand-soft text-brand",
  achieved: "bg-good-soft text-good",
  paused: "bg-warn-soft text-warn",
  dropped: "bg-line/50 text-muted",
};

export const STUDENT_LABELS: Record<string, string> = {
  active: "مستمرّ",
  paused: "متوقّف مؤقتاً",
  graduated: "تخرّج",
  withdrawn: "انسحب",
};

export const RELATION_LABELS: Record<string, string> = {
  father: "الأب",
  mother: "الأم",
  grandparent: "جد/جدة",
  sibling: "أخ/أخت",
  guardian: "وصي",
};

export const TREND_LABELS: Record<string, string> = {
  improving: "متحسّن",
  steady: "ثابت",
  declining: "متراجع",
};

export const TREND_TONES: Record<string, string> = {
  improving: "bg-good-soft text-good",
  steady: "bg-line/60 text-muted",
  declining: "bg-bad-soft text-bad",
};

export const ROLE_LABELS: Record<string, string> = {
  admin: "الإدارة",
  specialist: "أخصائية",
  guardian: "ولي أمر",
};

File: mobile/lib/core/format.dart
import 'package:intl/intl.dart';

/// Dates and numbers, formatted once.
///
/// ---------------------------------------------------------------------------
/// THE SERVER'S CLOCK IS UTC (`api/config/app.php` → `'timezone' => 'UTC'`).
///
/// That produces two shapes of timestamp, and both are UTC:
///
///   "2026-08-20T12:30:00.000000Z"  — every Eloquent date cast
///   "2026-08-24 12:30:00"          — `toDateTimeString()` and raw columns
///                                    (`next_session`, `read_by`, `last_at`)
///
/// The second has no zone on it, and `DateTime.parse` would read it as the
/// phone's local time — three hours off in summer. [Fmt.at] reads both as UTC
/// and hands back local time, which is also exactly what the web dashboard
/// shows (`new Date(...)` in the browser), so the two clients agree on when a
/// session was.
///
/// Date-only columns (`birth_date`, `started_at`) arrive as UTC midnight and
/// are converted the same way before the date is taken. Slicing the first ten
/// characters would also work today — but not if the server's timezone is
/// ever set to Asia/Hebron, when they arrive as 21:00Z the day before.
/// ---------------------------------------------------------------------------
abstract final class Fmt {
  static final _dayMonth = DateFormat('d MMMM', 'ar');
  static final _dayMonthYear = DateFormat('d MMMM yyyy', 'ar');
  static final _weekday = DateFormat('EEEE', 'ar');
  static final _number = NumberFormat.decimalPattern('en');

  static final _naive = RegExp(r'^\d{4}-\d{2}-\d{2}[ T]\d{2}:\d{2}(:\d{2})?$');
  static final _dateOnly = RegExp(r'^\d{4}-\d{2}-\d{2}$');

  /// Any timestamp the API sends, as local time.
  static DateTime? at(String? value) {
    if (value == null || value.isEmpty) return null;

    if (_dateOnly.hasMatch(value)) {
      return DateTime.tryParse(value);
    }

    if (_naive.hasMatch(value)) {
      return DateTime.tryParse('${value.replaceFirst(' ', 'T')}Z')?.toLocal();
    }

    return DateTime.tryParse(value)?.toLocal();
  }

  /// Latin digits throughout. intl writes the ar locale in Arabic-Indic ones,
  /// and "٢٣ أغسطس" above "89%" reads as two half-finished decisions. The web
  /// dashboard writes Latin digits too.
  static String _latin(String value) {
    const arabicIndic = '٠١٢٣٤٥٦٧٨٩';
    final buffer = StringBuffer();
    for (final char in value.split('')) {
      final index = arabicIndic.indexOf(char);
      buffer.write(index == -1 ? char : '$index');
    }
    return buffer.toString();
  }

  /// "23 أغسطس 2026"
  static String date(String? value) {
    final d = at(value);
    return d == null ? '—' : _latin(_dayMonthYear.format(d));
  }

  /// "23 أغسطس"
  static String shortDate(String? value) {
    final d = at(value);
    return d == null ? '—' : _latin(_dayMonth.format(d));
  }

  /// "الأحد 23 أغسطس" — for a diary, where the weekday is what is scanned for.
  static String dayDate(String? value) {
    final d = at(value);
    return d == null ? '—' : _latin('${_weekday.format(d)} ${_dayMonth.format(d)}');
  }

  /// "10:30"
  static String time(String? value) {
    final d = at(value);
    if (d == null) return '—';
    return '${d.hour.toString().padLeft(2, '0')}:${d.minute.toString().padLeft(2, '0')}';
  }

  /// "23 أغسطس، 10:30"
  static String dateTime(String? value) {
    final d = at(value);
    return d == null ? '—' : '${shortDate(value)}، ${time(value)}';
  }

  /// "اليوم" / "أمس" / "بكرة" / the weekday and date.
  static String relativeDay(String? value) {
    final d = at(value);
    if (d == null) return '—';

    final now = DateTime.now();
    final days = DateTime(d.year, d.month, d.day).difference(DateTime(now.year, now.month, now.day)).inDays;

    return switch (days) {
      0 => 'اليوم',
      -1 => 'أمس',
      1 => 'بكرة',
      _ => dayDate(value),
    };
  }

  /// "قبل 3 أيام" / "بعد ساعتين".
  ///
  /// Arabic counts in three forms — one, two, many — and "قبل 2 يوم" is the
  /// kind of thing that makes software read as foreign.
  static String relative(String? value) {
    final d = at(value);
    if (d == null) return '—';

    final diff = d.difference(DateTime.now());
    final past = diff.isNegative;
    final minutes = diff.inMinutes.abs();

    String phrase(int n, String one, String two, String few, String many) {
      if (n == 1) return one;
      if (n == 2) return two;
      if (n <= 10) return '$n $few';
      return '$n $many';
    }

    final String unit;
    if (minutes < 1) {
      return 'الآن';
    } else if (minutes < 60) {
      unit = phrase(minutes, 'دقيقة', 'دقيقتين', 'دقائق', 'دقيقة');
    } else if (minutes < 1440) {
      unit = phrase((minutes / 60).round(), 'ساعة', 'ساعتين', 'ساعات', 'ساعة');
    } else if (minutes < 43200) {
      unit = phrase((minutes / 1440).round(), 'يوم', 'يومين', 'أيام', 'يوماً');
    } else {
      unit = phrase((minutes / 43200).round(), 'شهر', 'شهرين', 'شهور', 'شهراً');
    }

    return past ? 'قبل $unit' : 'بعد $unit';
  }

  static String number(num? value) => value == null ? '—' : _number.format(value);

  /*
   * A counted noun, in Arabic's four shapes rather than English's two.
   *
   * «٢٩ أيام» is what a two-branch plural produces and it is wrong; Arabic
   * counts 1 alone, 2 as a dual, 3–10 with the broken plural, and 11 upward
   * with the singular again — «٢٩ يوماً». Zero takes the last shape too.
   *
   * The number is dropped for one and two because the word already carries
   * it: «يوم» is one day and «يومان» is two, and printing the digit beside
   * them reads as a stutter.
   *
   *   counted(1,  'يوم', 'يومان', 'أيام', 'يوماً')  →  يوم
   *   counted(2,  …)                                 →  يومان
   *   counted(5,  …)                                 →  5 أيام
   *   counted(29, …)                                 →  29 يوماً
   */
  static String counted(int n, String one, String two, String few, String many) =>
      switch (n) {
        1 => one,
        2 => two,
        >= 3 && <= 10 => '${number(n)} $few',
        _ => '${number(n)} $many',
      };

  /// The four shapes of the words this app counts most.
  static String days(int n) => counted(n, 'يوم', 'يومان', 'أيام', 'يوماً');
  static String sessions(int n) => counted(n, 'جلسة', 'جلستان', 'جلسات', 'جلسة');
  static String reports(int n) => counted(n, 'تقرير', 'تقريران', 'تقارير', 'تقريراً');
  static String messages(int n) => counted(n, 'رسالة', 'رسالتان', 'رسائل', 'رسالة');
  static String children(int n) => counted(n, 'طفل', 'طفلان', 'أطفال', 'طفلاً');

  static String percent(num? value) => value == null ? '—' : '${value.round()}%';

  /// "2026-08-23", for a date field the API validates with `date`.
  static String isoDate(DateTime value) =>
      '${value.year.toString().padLeft(4, '0')}-'
      '${value.month.toString().padLeft(2, '0')}-'
      '${value.day.toString().padLeft(2, '0')}';

  /// A date column from the API as the value of a date input.
  static String? toDateInput(String? value) {
    final d = at(value);
    return d == null ? null : isoDate(d);
  }

  /// A local moment, written the way the server stores it: UTC, no zone.
  ///
  /// Carbon does not convert an offset on the way into a datetime column — it
  /// keeps the wall-clock digits and drops the zone — so sending "10:30+03:00"
  /// would be stored as 10:30 UTC and shown back as 13:30. Converting here is
  /// what makes a session booked for 10:30 come back as 10:30, on the phone
  /// and on the dashboard alike.
  static String toServerDateTime(DateTime local) {
    final u = local.toUtc();
    String two(int n) => n.toString().padLeft(2, '0');
    return '${isoDate(u)} ${two(u.hour)}:${two(u.minute)}:00';
  }

  static String fileSize(int bytes) {
    if (bytes < 1024) return '$bytes B';
    if (bytes < 1024 * 1024) return '${(bytes / 1024).round()} KB';
    return '${(bytes / (1024 * 1024)).toStringAsFixed(1)} MB';
  }
}

/// The wording, in one place — the same words the web dashboard uses
/// (`web/lib/format.ts`). Everything below is read by a worried parent on a
/// phone, and "ما إجا ع الجلسة" is a letter where "الحالة: غياب" is a case file.
abstract final class Labels {
  static const levels = {
    4: 'مستقل',
    3: 'تلميح لفظي',
    2: 'مساعدة جزئية',
    1: 'مساعدة كاملة',
    0: 'رفض / ما استجاب',
  };

  /// What each step means — so two specialists score the same child the same.
  static const levelHints = {
    4: 'عملها لحاله بدون أي تذكير.',
    3: 'عملها لحاله بعد ما ذكّرناه بالكلام أو بالإشارة.',
    2: 'بدأ لحاله واحتاج مساعدة بالنص، أو لمسة توجيه.',
    1: 'عملناها معه يداً بيد من أولها لآخرها.',
    0: 'ما قبل يجرّب، أو ما استجاب إطلاقاً.',
  };

  static String level(int? value) => value == null ? '—' : levels[value] ?? '—';

  static const sessionStatus = {
    'scheduled': 'مجدولة',
    'held': 'تمّت',
    'absent': 'ما إجا',
    'excused': 'غياب بعذر',
    'cancelled': 'ملغية',
  };

  static const sessionTone = {
    'scheduled': 'muted',
    'held': 'good',
    'absent': 'bad',
    'excused': 'warn',
    'cancelled': 'muted',
  };

  static const reportStatus = {'draft': 'مسوّدة', 'published': 'وصل الأهل'};
  static const reportTone = {'draft': 'warn', 'published': 'good'};

  static const moods = ['happy', 'calm', 'tired', 'agitated', 'crying', 'resistant'];

  static const mood = {
    'calm': 'هادئ',
    'happy': 'مبسوط',
    'tired': 'تعبان',
    'agitated': 'متوتّر',
    'crying': 'بيبكي',
    'resistant': 'رافض',
  };

  static const moodFace = {
    'calm': '😌',
    'happy': '😄',
    'tired': '😪',
    'agitated': '😣',
    'crying': '😢',
    'resistant': '😤',
  };

  static const specialties = ['speech', 'occupational', 'behavioral', 'special_ed'];

  static const specialty = {
    'speech': 'النطق واللغة',
    'occupational': 'العلاج الوظيفي',
    'behavioral': 'تعديل السلوك',
    'special_ed': 'التربية الخاصة',
  };

  static const specialtyFull = {
    'speech': 'النطق واللغة والتخاطب',
    'occupational': 'العلاج الوظيفي والتكامل الحسي',
    'behavioral': 'تعديل السلوك والتوحد',
    'special_ed': 'التربية الخاصة وصعوبات التعلّم',
  };

  static const goalStatus = {
    'active': 'شغّال',
    'achieved': 'تحقّق',
    'paused': 'موقوف',
    'dropped': 'أُلغي',
  };

  static const goalTone = {
    'active': 'brand',
    'achieved': 'good',
    'paused': 'warn',
    'dropped': 'muted',
  };

  static const studentStatus = {
    'active': 'مستمرّ',
    'paused': 'متوقّف مؤقتاً',
    'graduated': 'تخرّج',
    'withdrawn': 'انسحب',
  };

  static const studentTone = {
    'active': 'good',
    'paused': 'warn',
    'graduated': 'brand',
    'withdrawn': 'muted',
  };

  static const enrollmentStatus = {'active': 'شغّال', 'paused': 'موقوف مؤقتاً', 'ended': 'انتهى'};

  static const relation = {
    'father': 'الأب',
    'mother': 'الأم',
    'grandparent': 'جد/جدة',
    'sibling': 'أخ/أخت',
    'guardian': 'وصي',
  };

  static const trend = {'improving': 'متحسّن', 'steady': 'ثابت', 'declining': 'متراجع'};
  static const trendTone = {'improving': 'good', 'steady': 'muted', 'declining': 'bad'};

  static const role = {'admin': 'الإدارة', 'specialist': 'أخصائية', 'guardian': 'ولي أمر'};

  static const gender = {'male': 'ذكر', 'female': 'أنثى'};
}
